flow/display_list_utils.cc

```diff
@@ -317,126 +431,158 @@ void DisplayListBoundsCalculator::drawPicture(const sk_sp<SkPicture> picture,
   if (pic_matrix) {
     pic_matrix->mapRect(&bounds);
   }
-  if (with_save_layer) {
-    accumulateRect(bounds);
-  } else {
-    matrix().mapRect(&bounds);
-    accumulator_->accumulate(bounds);
-  }
+  accumulateRect(bounds, with_save_layer ? kIsNonGeometric : kIsUnfiltered);
 }
 void DisplayListBoundsCalculator::drawDisplayList(
     const sk_sp<DisplayList> display_list) {
-  accumulateRect(display_list->bounds());
+  accumulateRect(display_list->bounds(), kIsUnfiltered);
 }
 void DisplayListBoundsCalculator::drawTextBlob(const sk_sp<SkTextBlob> blob,
                                                SkScalar x,
                                                SkScalar y) {
-  accumulateRect(blob->bounds().makeOffset(x, y));
+  accumulateRect(blob->bounds().makeOffset(x, y), kIsFilledGeometry);
 }
 void DisplayListBoundsCalculator::drawShadow(const SkPath& path,
                                              const SkColor color,
                                              const SkScalar elevation,
-                                             bool occludes,
+                                             bool transparentOccluder,
                                              SkScalar dpr) {
-  accumulateRect(
-      PhysicalShapeLayer::ComputeShadowBounds(path, elevation, dpr, matrix()));
+  SkRect shadow_bounds =
+      PhysicalShapeLayer::ComputeShadowBounds(path, elevation, dpr, matrix());
+  accumulateRect(shadow_bounds, kIsUnfiltered);
 }
 
-void DisplayListBoundsCalculator::accumulateRect(const SkRect& rect,
-                                                 bool forceStroke) {
-  SkRect dstRect = rect;
-  const SkPaint& p = paint();
-  if (forceStroke) {
-    if (p.getStyle() == SkPaint::kFill_Style) {
-      setDrawStyle(SkPaint::kStroke_Style);
-    } else {
-      forceStroke = false;
+bool DisplayListBoundsCalculator::getFilteredBounds(SkRect& bounds,
+                                                    SkImageFilter* filter) {
+  if (filter) {
+    if (!filter->canComputeFastBounds()) {
+      return false;
     }
+    bounds = filter->computeFastBounds(bounds);
   }
-  if (p.canComputeFastBounds()) {
-    dstRect = p.computeFastBounds(rect, &dstRect);
-    matrix().mapRect(&dstRect);
-    accumulator_->accumulate(dstRect);
+  return true;
+}
+
+bool DisplayListBoundsCalculator::adjustBoundsForPaint(SkRect& bounds,
+                                                       int flags) {
+  if ((flags & kIsUnfiltered) != 0) {
+    FML_DCHECK(flags == kIsUnfiltered);
+    return true;
+  }
+
+  if ((flags & kIsAnyGeometryMask) != 0) {
+    if ((flags & kIsDrawnGeometry) != 0) {
+      FML_DCHECK((flags & (kIsFilledGeometry | kIsStrokedGeometry)) == 0);
+      flags |= style_flag_;
+    }
+
+    // Path effect occurs before stroking...
+    if (path_effect_) {
+      SkPathEffect::DashInfo info;
+      if (path_effect_->asADash(&info) == SkPathEffect::kDash_DashType) {
```

Comment:
I don't believe we ever use this effect. There was a feature request for it that was long denied, and I think finally closed as won't fix. Users are expected to draw dashed lines manually.

Reply:
Text uses it.

Reply:
Ahh haha I forgot that one. Ok.